### PR TITLE
Cache AnalyzerConfigOptionsResult instances

### PR DESCRIPTION
**Pull request by agocke**
A common case for editorconfig is for most of the source files to have the same
options specified, which means if we store options for each source file it's
likely we'll have many duplicate sets. This change adds a cache to AnalyzerConfigSet
to try to de-dup identical options dictionaries to a single shared instance.

### CODE REVIEW

File: src/Compilers/Core/CodeAnalysisTest/AnalyzerConfigTests.cs [hunk 1]
```diff
@@ -1455,6 +1455,40 @@ public void DiagnosticIdInstancesAreSharedBetweenMultipleTrees()
...
+            configs.Add(Parse(@"
+[*.cs]
+dotnet_diagnostic.cs000.severity = warning", "/.editorconfig"));
+
+            var options = GetAnalyzerConfigOptions(
+                new[] { "/a.cs", "/b.cs", "/c.cs" },
+                configs);
+            configs.Free();
+
+            Assert.True(
```
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/AnalyzerConfigTests.cs]: consider using `Assert.Same`

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs [hunk 2]
```diff
@@ -594,5 +595,30 @@ internal static Location FirstOrNone(this ImmutableArray<Location> items)
         {
             return items.IsEmpty ? Location.None : items[0];
         }
+
+        internal sealed class SequenceEqualComparer<T> : IEqualityComparer<ImmutableArray<T>>
+        {
+            public static SequenceEqualComparer<T> Instance { get; } = new SequenceEqualComparer<T>();
+
+            public bool Equals(ImmutableArray<T> x, ImmutableArray<T> y)
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs]: Why not `x.GetSpan().SequenceEquals(y.GetSpan())`

Comment (agocke) [reply src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs]: We're running on Desktop, so how much slower is Slow Span?

Comment (agocke) [reply src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs]: On second thought, it should still be worth it.

File: src/Compilers/Core/CodeAnalysisTest/AnalyzerConfigTests.cs (hunk 1 above)
Comment (agocke) [reply src/Compilers/Core/CodeAnalysisTest/AnalyzerConfigTests.cs]: I think this is clearer.

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs (hunk 2 above)
Comment (agocke) [reply src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs]: I forgot the other problem. That helper requires T to implement IEquatable, which AnalyzerConfig.Section doesn't, and probably shouldn't. There's an open issue on CoreFX to add APIs for passing an IEqualityComparer https://github.com/dotnet/corefx/issues/35962

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs [hunk 3]
```diff
@@ -594,5 +595,30 @@ internal static Location FirstOrNone(this ImmutableArray<Location> items)
         {
             return items.IsEmpty ? Location.None : items[0];
         }
+
+        internal sealed class SequenceEqualComparer<T> : IEqualityComparer<ImmutableArray<T>>
```
Comment (cston) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs]: >SequenceEqualComparer [](start = 30, length = 21)

Why a nested class?

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs [hunk 4]
```diff
@@ -594,5 +595,30 @@ internal static Location FirstOrNone(this ImmutableArray<Location> items)
...
+            public bool Equals(ImmutableArray<T> x, ImmutableArray<T> y)
+            {
+                if (x.Length != y.Length)
+                {
+                    return false;
+                }
+
+                for (int i = 0; i < x.Length; i++)
+                {
+                    if (!x[i].Equals(y[i]))
```
[the remark this answers is not in the document]
Comment (agocke) [reply src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs]: This is internal, so I always follow YAGNI

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs (hunk 3 above)
Comment (agocke) [reply src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs]: It's pretty niche. We haven't needed it for the past 7 years. Why not hide it from completion?

Comment (cston) [reply src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs]: In that case, consider making it `private` to `AnalyzerConfigSet`.

---
In reply to: [327737635](https://github.com/dotnet/roslyn/pull/38760#discussion_r327737635) [](ancestors = 327737635)

Comment (agocke) [reply src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs]: Yeah I think that's a good idea. Will-do.

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs [hunk 5]
```diff
@@ -143,15 +153,30 @@ public AnalyzerConfigOptionsResult GetOptionsForSourcePath(string sourcePath)
...
                 }
             }
 
-            return new AnalyzerConfigOptionsResult(
-                treeOptionsBuilder.Count > 0 ? treeOptionsBuilder.ToImmutable() : SyntaxTree.EmptyDiagnosticOptions,
-                analyzerOptionsBuilder.Count > 0 ? analyzerOptionsBuilder.ToImmutable() : AnalyzerConfigOptions.EmptyDictionary,
-                diagnosticBuilder.ToImmutableAndFree());
+            // Try to avoid creating extra dictionaries if we've already seen an options result with the
+            // exact same options
+            var sectionKey = sectionBuilder.ToImmutableAndFree();
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs]: Using an `ImmutableArray<Section>` here means that we are at minimum allocating an array for every lookup attempt. That's even in the good case where we have a cache hit and all the builders return optimal values. Why not instead use `List<Section>` as the key? This is a completely private cache hence it's relatively easy to guard against mutations. That would also mean in the optimal case the lookup is allocation free because we'll get back a `List<Section>` that already has sufficient size to build up the key.

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs [hunk 6]
```diff
@@ -37,6 +38,16 @@ public sealed class AnalyzerConfigSet
         private readonly ConcurrentDictionary<ReadOnlyMemory<char>, string> _diagnosticIdCache =
             new ConcurrentDictionary<ReadOnlyMemory<char>, string>(CharMemoryEqualityComparer.Instance);
 
+        // PERF: Most files will probably have the same options, so share the dictionary instances
+        private readonly ConcurrentCache<ImmutableArray<Section>, AnalyzerConfigOptionsResult> _optionsCache =
+            new ConcurrentCache<ImmutableArray<Section>, AnalyzerConfigOptionsResult>(50, SequenceEqualComparer<Section>.Instance); // arbitrary size
+
+        private readonly ObjectPool<TreeOptions.Builder> _treeOptionsPool =
+            new ObjectPool<TreeOptions.Builder>(() => ImmutableDictionary.CreateBuilder<string, ReportDiagnostic>(CaseInsensitiveComparison.Comparer));
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs]: Why this comparer vs. the named comparers that already exist in this section of the code?

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs [hunk 7]
```diff
@@ -37,6 +38,16 @@ public sealed class AnalyzerConfigSet
...
+
+        private readonly ObjectPool<AnalyzerOptions.Builder> _analyzerOptionsPool =
+            new ObjectPool<AnalyzerOptions.Builder>(() => ImmutableDictionary.CreateBuilder<string, string>(CaseInsensitiveComparison.Comparer));
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs]: Why this comparer vs. the named comparers that already exist in this section of the code?

**Comment by agocke** [comment]
Effect of changes:

Before:

```
|               Method |     Mean |    Error |   StdDev |   Median |      Min |      Max |     Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------|---------:|---------:|---------:|---------:|---------:|---------:|----------:|------:|------:|----------:|                          
| BuildAnalyzerConfigs | 99.95 ms | 3.834 ms | 11.25 ms | 97.49 ms | 84.06 ms | 128.4 ms | 1000.0000 |     - |     - |   8.37 MB |
```

After:

```
|               Method |     Mean |    Error |   StdDev |   Median |      Min |      Max |     Gen 0 | Gen 1 | Gen 2 | Allocated |                          
|----------------------|---------:|---------:|---------:|---------:|---------:|---------:|----------:|------:|------:|----------:|                          
| BuildAnalyzerConfigs | 96.93 ms | 3.709 ms | 10.46 ms | 93.67 ms | 83.76 ms | 127.1 ms | 1000.0000 |     - |     - |   8.36 MB |
```

File: src/Compilers/Core/CodeAnalysisTest/AnalyzerConfigTests.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/Core/CodeAnalysisTest/AnalyzerConfigTests.cs]: I disagree. `Assert.Same` exists for this reason, and will give a better error message if it fails.

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs [hunk 8]
```diff
@@ -10,6 +10,7 @@
 using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
```
Comment (333fred) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs]: Unnecessary change?

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs [hunk 9]
```diff
@@ -143,15 +180,31 @@ public AnalyzerConfigOptionsResult GetOptionsForSourcePath(string sourcePath)
...
-                diagnosticBuilder.ToImmutableAndFree());
+            // Try to avoid creating extra dictionaries if we've already seen an options result with the
+            // exact same options
+            if (!_optionsCache.TryGetValue(sectionKey, out var result))
+            {
+                result = new AnalyzerConfigOptionsResult(
+                    treeOptionsBuilder.Count > 0 ? treeOptionsBuilder.ToImmutable() : SyntaxTree.EmptyDiagnosticOptions,
+                    analyzerOptionsBuilder.Count > 0 ? analyzerOptionsBuilder.ToImmutable() : AnalyzerConfigOptions.EmptyDictionary,
+                    diagnosticBuilder.ToImmutableAndFree());
+                _optionsCache.TryAdd(sectionKey, result);
```
Comment (333fred) [inline src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs]: Should we throwing away result and getting from the cache if TryAdd fails so that GC can free this `AnalyzerConfigOptionsResult` while it's still cheap? If not, consider adding a discard assignment to be clear about intent.

**Comment by 333fred** [comment]
Done review pass (commit 2)

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs (hunk 9 above)
Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs]: I don't think it really matters either way, races should be rare.

Comment (333fred) [reply src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs]: I'm not talking about races. This change is about reducing memory pressure, and this will keep an extra class around for a while. If we were to immediately remove the reference and use the one in the cache, the GC pass would be basically free as it's still in Gen 0.

Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs]: Yeah, I'm saying that it should be rare for that to happen. Most calls will hit the cache, so you may end up with one or two extra objects, but that's not what the problem is about. Note that the cache is not perfect either, there's a chance you could end up evicting dictionaries that could be shared.

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs [hunk 10]
```diff
@@ -37,6 +37,43 @@ public sealed class AnalyzerConfigSet
...
+
+                for (int i = 0; i < x.Count; i++)
+                {
+                    if (!ReferenceEquals(x[i], y[i]))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
```
Comment (cston) [inline src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs]: Can this be implemented with `System.Linq.Enumerable.SequenceEqual()`?
```
return x.SequenceEqual(y);
```

Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs]: No constant-time easy-out for incorrect lengths.

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs (hunk 9 above)
Comment (333fred) [reply src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs]: Alright, fair enough.
